learn-arai/app/hooks/useAuth.ts

```diff
@@ -0,0 +1,79 @@
+'use client';
+
+// this react hook responsible for checking if user is logged in.
+import { redirect } from 'next/navigation';
+
+import { useEffect, useState } from 'react';
+
+import { User } from './useUser';
+import { useUser } from './useUser';
+
+export const useAuth = () => {
+    const { addUser, removeUser, user, fetchedUser, getUserFromLocalStorage } =
+        useUser();
+    const [isAuthenticated, setIsAuthenticated] = useState<boolean>();
+
+    useEffect(() => {
+        checkSession();
+    });
+
+    const signIn = async (credentials: FormData) => {
+        const response = await sendCredentialToServer(credentials);
+        const email = await fetchedUser();
+        addUser(email);
+
+        return response;
+    };
+
+    const signOut = () => {
+        removeUser();
+    };
+
+    const sendCredentialToServer = async (formData: FormData) => {
+        const response = await fetch('http://localhost:3000/auth/sign-in', {
+            method: 'POST',
+            body: formData,
+            credentials: 'include',
+        });
+
+        const data = await response.json();
+        const message = data.message;
+        const status = data.status;
+
+        return { message, status };
+    };
+
+    const checkSession = async () => {
+        const response = await fetch(
+            'http://localhost:3000/auth/session-check',
+            {
+                method: 'GET',
+                credentials: 'include',
+            }
+        );
+
+        const data = await response.json();
+        const isSessionExpire = data.isSessionExpire;
+
+        const currentPath = window.location.pathname;
+        if (isSessionExpire && currentPath != '/login')
+            window.location.href = '/login';
+
+        // if session is not expire then isSession = `false` and isAuthenticated
+        // will set to `true`
+        setIsAuthenticated(!isSessionExpire);
+
+        if (!isSessionExpire) {
+            await getUserFromLocalStorage();
+        }
```

Comment:
If the session is expired just set user object to null this will also leads to invalid state